third_party/pcsc-lite/naclport/server_clients_management/src/server-request-handler.js

```diff
@@ -0,0 +1,336 @@
+  this.id = idGenerator.next();
+
+  const nameForJsLog =
+      this.clientNameForLog_ ? this.clientNameForLog_ : 'ourselves';
+  /**
+   * @type {!goog.log.Logger}
+   * @const
+   */
+  this.logger = GSC.Logging.getScopedLogger(
+      `Pcsc.ServerRequestHandler<${nameForJsLog}>, id=${this.id}>`);
+
+  /** @private */
+  this.serverMessageChannel_ = serverMessageChannel;
+
+  /** @private */
+  this.serverReadinessTracker_ = serverReadinessTracker;
+
+
+  /**
+   * @type {GSC.Requester?}
+   * @private
+   */
+  this.serverRequester_ = null;
+
+  // The requests processing is deferred until the server readiness is
+  // established.
+  /** @private */
+  this.deferredProcessor_ =
+      new DeferredProcessor(this.serverReadinessTracker_.promise);
+
+  this.serverMessageChannel_.addOnDisposeCallback(
+      this.serverMessageChannelDisposedListener_.bind(this));
+};
+
+const ServerRequestHandler =
+    GSC.PcscLiteServerClientsManagement.ServerRequestHandler;
+
+goog.inherits(ServerRequestHandler, goog.Disposable);
+
+/**
+ * Generator that is used by the ServerRequestHandler class to generate unique
+ * client handler identifiers.
+ *
+ * This is a singleton object, because the identifiers have to be unique among
+ * all client handler instances.
+ * @type {!goog.iter.Iterator.<number>}
+ */
+const idGenerator = goog.iter.count();
+
+/** @override */
+ServerRequestHandler.prototype.disposeInternal = function() {
+  ServerRequestHandler.base(this, 'disposeInternal');
+
+
+  if (this.serverRequester_) {
+    this.serverRequester_.dispose();
+    this.serverRequester_ = null;
+
+    // Having the this.serverRequester_ member initialized means that the
+    // message to add a new client handler has been sent to the server. This is
+    // the right place to send to the server the message to delete the client
+    // handler.
+    //
+    // However, there is an additional condition checked here, which skips
+    // sending the message to the server if this.serverMessageChannel_ is null.
+    // That's because the reason of client handler disposal may be the disposal
+    // of the server message channel, in which case this.serverMessageChannel_
+    // will be cleared before reaching this point (see the
+    // serverMessageChannelDisposedListener_ method).
+    if (this.serverMessageChannel_)
+      this.sendServerDeleteHandlerMessage_();
+  }
+
+  this.serverReadinessTracker_ = null;
+
+  // Note that this statement has to be executed after the
+  // sendServerDeleteHandlerMessage_ method is called, because the latter
+  // accesses this member.
+  this.serverMessageChannel_ = null;
+};
+
+/**
+ * @param {!RemoteCallMessage} remoteCallMessage The remote call message
+ * containing the request contents.
+ * @return {!goog.Promise}
+ */
+ServerRequestHandler.prototype.handleRequest =
+    function(remoteCallMessage) {
+  return this.deferredProcessor_.addJob(
+      this.postRequestToServer_.bind(this, remoteCallMessage));
+}
+
+    /**
```

Comment:
nit: The indentation is messed up here (likely because of a missing semicolon above).